druid/src/scroll_component.rs

```diff
@@ -0,0 +1,456 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A component for embedding in another widget to provide consistant and
+//! extendable scrolling behavior
+
+use std::time::Duration;
+
+use crate::kurbo::{Affine, Point, Rect, RoundedRect, Size, Vec2};
+use crate::theme;
+use crate::{
+    Env, Event, EventCtx, LifeCycle, LifeCycleCtx, PaintCtx, Region, RenderContext, TimerToken,
+};
+
+//TODO: Add this to env
+pub const SCROLLBAR_MIN_SIZE: f64 = 45.0;
+
+#[derive(Debug, Copy, Clone)]
+pub enum BarHoveredState {
+    None,
+    Vertical,
+    Horizontal,
+}
+
+impl BarHoveredState {
+    pub fn is_hovered(self) -> bool {
+        matches!(
+            self,
+            BarHoveredState::Vertical | BarHoveredState::Horizontal
+        )
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum BarHeldState {
+    None,
+    /// Vertical scrollbar is being dragged. Contains an `f64` with
+    /// the initial y-offset of the dragging input
+    Vertical(f64),
+    /// Horizontal scrollbar is being dragged. Contains an `f64` with
+    /// the initial x-offset of the dragging input
+    Horizontal(f64),
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct ScrollbarsState {
+    pub opacity: f64,
+    pub timer_id: TimerToken,
+    pub hovered: BarHoveredState,
+    pub held: BarHeldState,
+}
+
+impl Default for ScrollbarsState {
+    fn default() -> Self {
+        Self {
+            opacity: 0.0,
+            timer_id: TimerToken::INVALID,
+            hovered: BarHoveredState::None,
+            held: BarHeldState::None,
+        }
+    }
+}
+
+impl ScrollbarsState {
+    /// true if either scrollbar is currently held down/being dragged
+    pub fn are_held(&self) -> bool {
+        !matches!(self.held, BarHeldState::None)
+    }
+}
+
+/// Embedable component exposing reusable scroll handling logic.
+///
+/// In most situations composing [`Scroll`] or [`List`] is a better idea
+/// for general UI construction. However some cases are not covered by
+/// composing those widgets, such as when a widget needs fine grained
+/// control over its scrolling state or doesn't make sense to exist alone
+/// without scrolling behavior.
+///
+/// `ScrollComponent` contains the unified and consistant scroll logic
+/// used by both [`Scroll`] and [`List`]. This can be used to add this
+/// logic to a custom widget when the need arises.
+///
+/// To use, instance in your widget's new fn and place in a field, keep
+/// the [`content_size`] field updated as scrollable content size changes,
+/// call [`event`] and [`lifecycle`] with all event and lifecycle events,
+/// call [`handle_scroll`] with all events after all other event handling
+/// code and finally perform painting from within a closure provided to
+/// [`paint_content`].
```

Comment:
This would be my attempt at explaining how to use it:
```suggestion
/// It should be used like this:
/// - Store an instance of `ScrollComponent` in your widget's struct.
/// - During layout, set the [`content_size`] field to the child's size.
/// - Call [`event`] and [`lifecycle`] with all event and lifecycle events before propagating them to children.
/// - Call [`handle_scroll`] with all events after handling / propagating them.
/// - And finally perform painting using the provided [`paint_content`] function.
///
/// Also, taking a look at the [`Scroll`] source code can be helpful.
```

Reply:
That is a very top notch explanation :+1: